pkg/controller/spec_processor.go

```diff
@@ -122,18 +131,37 @@ func (sp *SpecProcessor) processMatch(selector string, primitivesOnly bool) (int
 	if _, allowed := sp.allowedResources[objName]; !allowed {
 		return nil, errors.Errorf("references can only point at direct dependencies: %s", selector)
 	}
+
+	var objToTraverse interface{}
+	switch resourceOutputName {
+	case "":
+		objToTraverse = resInfo.actual.Object
+	case ResourceOutputNameBindSecret:
+		if resInfo.serviceBindingSecret == nil {
+			return nil, errors.Errorf("%q requested, but %q is not a ServiceBinding", resourceOutputName, objName)
+		}
+		objToTraverse = resInfo.serviceBindingSecret.Data
+	default:
+		return nil, errors.Errorf("resource output name %q not understood for %q", resourceOutputName, objName)
+	}
+
 	// To avoid overcomplicated format of reference like this: {{{res1#{$.a.string}}}}
 	// And have something like this instead: {{{res1#a.string}}}
 	jsonPath := fmt.Sprintf("{$.%s}", parts[1])
-	fieldValue, err := resources.GetJsonPathValue(resInfo.actual.Object, jsonPath, false)
+	fieldValue, err := resources.GetJsonPathValue(objToTraverse, jsonPath, false)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to process JsonPath reference %s", selector)
 	}
 	if fieldValue == nil {
 		return nil, errors.Errorf("field not found: %s", selector)
 	}
+
 	if primitivesOnly {
-		switch fieldValue.(type) {
+		switch typedFieldValue := fieldValue.(type) {
+		case []byte:
+			// Secrets are in bytes. We wildly cast them to a string and hope for the best
+			// so we can put them in the JSON in a 'nice' way.
```

Comment:
What will happen if the data is not a valid utf8 sequence? I think json encoder will replace "invalid" bytes with some special character. We can catch it here properly (i.e. check for utf8 validity; there must be a func for this somewhere) and return an error.

Reply:
Indeed, I was lazy. Will fix, that is definitely better behaviour.

Reply:
The comment does not to be here anymore?

Reply:
I left it there because it's still a bit rude to cast a byte array to a json string just because it works. Might be an accident.